### ISSUE TITLE
Local Attention vs Standard when length < window_size

### ISSUE THREAD

**Issue by gautierdag** [lucidrains/local-attention#4]
Hi,
I'm having trouble finding the difference in the Local Attention implemented in this repo and the standard global mechanism when the window should cover the entire sequence.

I've tried implementing a Multihead Local Attention module similar to the standard PyTorch (`nn.MultiheadAttention`) and I was expecting that for when the local attention `window_size` is always less than the sequence lengths, I would obtain the same results as the global attention mechanism (with causal attention mask).

```python
class LocalMultiheadAttention(nn.Module):
    def __init__(
        self, embed_dim=64, num_heads=4, local_window_size=100, dropout=0.0,
    ):
        super(LocalMultiheadAttention, self).__init__()
        self.attention = LocalAttention(
            window_size=local_window_size,  # window size. 512 is optimal, but 256 or 128 yields good enough results
            causal=True,  # auto-regressive or not
            look_backward=1,  # each window looks at the window before
            dropout=dropout,  # post-attention dropout
            exact_windowsize=False,  # if this is set to true, in the causal setting, each query will see at maximum the number of keys equal to the window size
            autopad=True,
        )
        self.wq = nn.Linear(embed_dim, embed_dim, bias=False)
        self.wk = nn.Linear(embed_dim, embed_dim, bias=False)
        self.wv = nn.Linear(embed_dim, embed_dim, bias=False)
        self.wo = nn.Linear(embed_dim, embed_dim, bias=True)
        self.n_heads = num_heads

        assert (
            embed_dim // num_heads
        ) * num_heads == embed_dim, "embed_dim must be divisible by num_heads"

    def forward(self, query, key, value, **kwargs):
        # handle seq first (L X B X E)
        query = query.permute(1, 0, 2)
        key = key.permute(1, 0, 2)
        value = value.permute(1, 0, 2)
        query, key, value = self.wq(query), self.wk(key), self.wv(value)
        query, key, value = (x.chunk(self.n_heads, -1) for x in (query, key, value))
        return self.wo(
            torch.cat(
                [self.attention(q, k, v) for q, k, v in zip(query, key, value)], -1
            )
        ).permute(1, 0, 2)
```

However, I find that the local attention does not converge at all compared to the global version even though the local attention window > sequence length, and so it should act the same as the global attention. I can't figure out why that could be the case, and would really appreciate any explanation, comment or help 😅

**Comment by lucidrains** [comment]
@gautierdag Hi Gautier! The way you have it, each head will only have a dimension of 16, which is way too small to learn anything. You want to keep the head dimension at 32 at least, and 64 preferrable

**Comment by gautierdag** [comment]
Thank you for your comment and your many repos 🙂! 

Unfortunately, the problem is not in the dimensions since the Pytorch multi-head attention module (`nn.MultiheadAttention`) also splits the embedding into distinct parts under the hood, and it works fine with the same dimensions. It might be something else on my end..

I've dug more and compared a wide range of different inputs and indeed it works as expected, so I have a feeling my bug is in the way I'm dealing with self-attention vs decoder attention - which is outside the scope of this.

Thank again!
